rmw_dds_common/src/graph_cache.cpp

```diff
@@ -0,0 +1,962 @@
+        rmw_ret = RMW_RET_BAD_ALLOC;
+        goto cleanup;
+      }
+      topic_names_and_types->types[index].data[type_index] = type_name;
+      ++type_index;
+    }
+    ++index;
+  }
+  return RMW_RET_OK;
+cleanup:
+  if (RMW_RET_OK != rmw_names_and_types_fini(topic_names_and_types)) {
+    RCUTILS_LOG_ERROR_NAMED(
+      log_tag,
+      "error during report of error: %s", rmw_get_error_string().str);
+  }
+  return rmw_ret;
+}
+
+rmw_ret_t
+GraphCache::get_names_and_types(
+  DemangleFunctionT demangle_topic,
+  DemangleFunctionT demangle_type,
+  rcutils_allocator_t * allocator,
+  rmw_names_and_types_t * topic_names_and_types) const
+{
+  assert(demangle_topic);
+  assert(demangle_type);
+  RCUTILS_CHECK_ALLOCATOR_WITH_MSG(allocator, "get_node_names allocator is not valid",
+    return RMW_RET_INVALID_ARGUMENT);
+  if (RMW_RET_OK != rmw_names_and_types_check_zero(topic_names_and_types)) {
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+
+  NamesAndTypes topics;
+  {
+    std::lock_guard<std::mutex> guard(mutex_);
+    __get_names_and_types(
+      data_readers_,
+      demangle_topic,
+      demangle_type,
+      topics);
+    __get_names_and_types(
+      data_writers_,
+      demangle_topic,
+      demangle_type,
+      topics);
+  }
+
+  return __populate_rmw_names_and_types(
+    topics,
+    allocator,
+    topic_names_and_types);
+}
+
+static
+const rmw_dds_common::msg::NodeEntitiesInfo *
+__find_node(
+  const GraphCache::ParticipantToNodesMap & participant_map,
+  const std::string & node_name,
+  const std::string & node_namespace)
+{
+  for (const auto & participant : participant_map) {
+    for (const auto & node : participant.second) {
+      if (
+        node.node_name == node_name &&
+        node.node_namespace == node_namespace)
+      {
+        return &node;
+      }
+    }
+  }
+  return nullptr;
+}
+
+static
+NamesAndTypes
+__get_names_and_types_from_gids(
+  const GraphCache::EntityGidToInfo & entities_map,
+  const GraphCache::GidSeq & gids,
+  DemangleFunctionT demangle_topic,
+  DemangleFunctionT demangle_type)
+{
+  NamesAndTypes topics;
+
+  for (const auto & gid_msg : gids) {
+    rmw_gid_t gid;
+    rmw_dds_common::convert_msg_to_gid(&gid_msg, &gid);
+    auto it = entities_map.find(gid);
+    if (it == entities_map.end()) {
+      continue;
+    }
+    std::string demangled_topic_name = demangle_topic(it->second.topic_name);
+    if ("" == demangled_topic_name) {
+      continue;
+    }
+    topics[demangled_topic_name].insert(demangle_type(it->second.topic_type));
+  }
+  return topics;
+}
+
+static
+rmw_ret_t
+__get_names_and_types_by_node(
+  const GraphCache::ParticipantToNodesMap & participants_map,
+  const GraphCache::EntityGidToInfo & entities_map,
+  const std::string & node_name,
+  const std::string & namespace_,
+  DemangleFunctionT demangle_topic,
+  DemangleFunctionT demangle_type,
+  const GraphCache::GidSeq & (*get_entities_gids)(const rmw_dds_common::msg::NodeEntitiesInfo &),
```

Comment:
@ivanpauno nit: for improved readability, consider aliasing this type like you do for `DemangleFunctionT`